Unify boxed function signature between jit and c10

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37034 Unify boxed function signature between jit and c10**

c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

smessmer committed:
Unify boxed function signature between jit and c10

c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/stack.h:
```diff
@@ -72,11 +87,14 @@ template <typename... Types>
   size_t i = 0;
   constexpr size_t N = sizeof...(args);
-  int result[N] = {
+  (void)std::initializer_list<int>{
```
This change isn't strictly necessary, right?

smessmer replied:
true, this is something I cleaned up on the way

smessmer replied:
@bhosmer lol remember the msvc issue where we had to land exactly this to fix it? I should have landed this sooner and we would have had to do a lot less debugging there...

bhosmer replied:
ugh yeah as you were describing the fix I was thinking of this change and wondering where it had gone 😝

bhosmer reviewed (approved):
LGTM

smessmer commented on aten/src/ATen/core/stack.h:
```diff
@@ -29,9 +29,15 @@ using Operation = std::function<int(Stack&)>;
 }
+static inline IValue& peek(Stack* stack, size_t i, size_t N) {
+  return peek(*stack, i, N);
+}
```
because otherwise we'd have to change even more code. They probably should die at some point though.

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

smessmer committed:
Update on "Unify boxed function signature between jit and c10"


c10 takes a Stack* in boxed functions while JIT took Stack&.
c10 doesn't return anything while JIT returns an int which is always zero.

This changes JIT to follow the c10 behavior.

Differential Revision: [D20567950](https://our.internmc.facebook.com/intern/diff/D20567950/)

[ghstack-poisoned]

bddppq commented:
Our project has registered custom jit operators (outside of the pytorch tree), and this diff breaks them because the `Operation` signature has been changed.
Is it that directly registering jit operator considered jit internal details and so not much backward compatibility guarantee going forward? From quickly browsing some projects that adds custom op, they seem to directly register jit operator as well (and this change actually also breaks them)
- https://github.com/NVIDIA/TRTorch/blob/223d8ddf9301a3fcb68167757a3f997b9e0c7d45/core/execution/register_trt_op.cpp#L58
- https://github.com/pytorch/tvm/blob/9d3ca577133152b04820fdd283291df5aacc6415/torch_tvm/register.h#L38

smessmer commented:
The `torch::jit::RegisterOperators` API is only meant for internal prim ops and not public API. Does the [public operator registration API](https://pytorch.org/tutorials/advanced/torch_script_custom_ops.html) work for your use case?

smessmer commented:
Btw, are you certain those projects break? I think TVM will likely break because they need access to the JIT node and that changed, but I think the TRTorch project shouldn't break. The way they're calling still works, just throws a deprecation warning now.

bddppq commented:
The reason we chose to directly register jit operator is that our custom op's kernel needs to access the attribute of its associated jit node (similar to the torch_tvm use case linked in my last comment).

bddppq commented:
> Btw, are you certain those projects break? I think TVM will likely break because they need access to the JIT node and that changed, but I think the TRTorch project shouldn't break. The way they're calling still works, just throws a deprecation warning now.
I have not tried actually building them, maybe you are right that TRTorch doesn't break.

smessmer commented:
hm yes, in that case you need `torch::jit::RegisterOperators`. Note that the whole jit Node class isn't part of the public API and could change. It's ok to use it but you'll likely have to adapt your code from time to time.
